Implement get_backend_info()

HPS-1 opened the pull request.

AlexeySachkov commented on sycl/source/detail/context_impl.cpp:
```diff
@@ -242,6 +242,40 @@ context_impl::get_info<info::context::atomic_fence_scope_capabilities>() const {
 }
 
+template <>
+std::string context_impl::get_backend_info<info::platform::version>() const {
```
Should we also use some traits to encode return type instead of hardcoding it here? This way if it is ever changed, it would be enough to change it just once in a single place

AlexeySachkov reviewed (commented):
Tests are missing. We should at least add some to `sycl/test` which would check that we can compile  code using new API and/or there could be tests in `sycl/test-e2e` which would check that the new API behaves as expected

HPS-1 commented:
> Tests are missing. We should at least add some to `sycl/test` which would check that we can compile code using new API and/or there could be tests in `sycl/test-e2e` which would check that the new API behaves as expected

Fixed! Added the new e2e test case sycl/test-e2e/backend_info.cpp for this purpose.

againull reviewed (commented):
@HPS-1  

> One known concern: it seems that sycl::platform, sycl::context and sycl::kernel may have multiple associated device, but according to the spec the return type for sycl::xxx::get_backend_info<info::device::version>() should be std::string (i.e. a single device version) so I'm just returning the version of the first associated device in the list. Is this OK?

`info::device::version` is the information descriptor of the `sycl::device`. Do we have to actually support it for all other classes `sycl::platform`, `sycl::context` and `sycl::kernel`? Maybe it's enough to support it only for sycl::device.

againull commented on sycl/source/detail/event_impl.cpp:
```diff
@@ -407,6 +407,60 @@ event_impl::get_info<info::event::command_execution_status>() {
+template <>
+typename info::platform::version::return_type
+event_impl::get_backend_info<info::platform::version>() const {
+  if (MContext->getBackend() != backend::opencl) {
```
AFAIK  Event may have uninitialized context, context is initialized lazily for the event. For example, default-constructed event doesn't have context.

HPS-1 replied:
Thanks for the reminder! I've changed this to use getContextImplPtr() which will initialize a default context instead of using MContext directly.

againull commented on sycl/include/sycl/info/sycl_backend_traits.def:
```diff
@@ -0,0 +1,3 @@
+__SYCL_PARAM_TRAITS_SPEC(platform, version, std::string, PI_PLATFORM_INFO_VERSION)
+__SYCL_PARAM_TRAITS_SPEC(device, version, std::string, PI_DEVICE_INFO_VERSION)
+__SYCL_PARAM_TRAITS_SPEC(device, backend_version, std::string, PI_DEVICE_INFO_BACKEND_VERSION)
```
Please fix "No newline at end of file"

againull commented on sycl/source/detail/event_impl.cpp:
```diff
@@ -407,6 +407,60 @@ event_impl::get_info<info::event::command_execution_status>() {
+  if (MContext->getBackend() != backend::ext_oneapi_level_zero) {
+    throw sycl::exception(errc::backend_mismatch,
+                          "the info::device::backend_version info descriptor "
+                          "can only be queried with a level0 backend");
```
If I am not mistaken now we use "Level Zero" everywhere instead of level0 or L0 or other names.

HPS-1 commented:
> @HPS-1
> 
> > One known concern: it seems that sycl::platform, sycl::context and sycl::kernel may have multiple associated device, but according to the spec the return type for sycl::xxx::get_backend_infoinfo::device::version() should be std::string (i.e. a single device version) so I'm just returning the version of the first associated device in the list. Is this OK?
> 
> `info::device::version` is the information descriptor of the `sycl::device`. Do we have to actually support it for all other classes `sycl::platform`, `sycl::context` and `sycl::kernel`? Maybe it's enough to support it only for sycl::device.

@againull Hi! Thanks for reviewing the code. And about the question you have, the thing is that the SYCL specification for (all variations of) get_backend_info() says "The type alias Param::return_type must be defined in accordance with the SYCL backend specification.", while the SYCL backend spec (for OpenCL, link here https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_backend_specific_information_descriptors) suggest that both info::platform::version and info::device::version are backend-specific info descriptors. I think this means that we should allow the user to query them using all variations of get_backend_info(), including the ones you mentioned such as sycl::context::get_backend_info().

againull commented:
> > @HPS-1
> > > One known concern: it seems that sycl::platform, sycl::context and sycl::kernel may have multiple associated device, but according to the spec the return type for sycl::xxx::get_backend_infoinfo::device::version() should be std::string (i.e. a single device version) so I'm just returning the version of the first associated device in the list. Is this OK?
> > 
> > 
> > `info::device::version` is the information descriptor of the `sycl::device`. Do we have to actually support it for all other classes `sycl::platform`, `sycl::context` and `sycl::kernel`? Maybe it's enough to support it only for sycl::device.
> 
> @againull Hi! Thanks for reviewing the code. And about the question you have, the thing is that the SYCL specification for (all variations of) get_backend_info() says "The type alias Param::return_type must be defined in accordance with the SYCL backend specification.", while the SYCL backend spec (for OpenCL, link here https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_backend_specific_information_descriptors) suggest that both info::platform::version and info::device::version are backend-specific info descriptors. I think this means that we should allow the user to query them using all variations of get_backend_info(), including the ones you mentioned such as sycl::context::get_backend_info().


In such case probably we can return the version of the default/selected device.
For example, let's say if we have 3 devices (opencl gpu, l0 gpu, opencl cpu) in the context.
If user didn't set ONEAPI_DEVICE_SELECTOR, then we can return version of the device chosen by default selector - it will be l0 gpu.
If user set ONEAPI_DEVICE_SELECTOR=opencl:cpu, then we can return the version of the opencl cpu device.

againull commented on sycl/source/detail/event_impl.cpp:
```diff
@@ -407,6 +407,69 @@ event_impl::get_info<info::event::command_execution_status>() {
+      return Queue->getDeviceImplPtr()
+          ->get_platform()
+          .get_info<info::platform::version>();
+    } catch (...) {
+      std::rethrow_exception(std::current_exception());
+    }
```
Could you please clarify what is the point of doing this?

againull replied:
Discussed offline, this is redundant, so needs to be removed. Approving the PR on condition that this will be fixed before merging.
